src/core/unit.h

```diff
@@ -110,6 +110,72 @@ typedef union UnitDependencyInfo {
         } _packed_;
 } UnitDependencyInfo;
 
+/* Store information about why a unit was activated.
+ * We start with trigger units (.path/.timer), eventually it will be expanded to include more metadata. */
+typedef struct ActivationEventInfo {
+        UnitType trigger_unit_type;
+        unsigned n_ref;
+        char *trigger_unit_name;
+} ActivationEventInfo;
+
+/* For casting an activation event into the various unit-specific types */
+#define DEFINE_EVENT_INFO_CAST(UPPERCASE, MixedCase, UNIT_TYPE)                 \
+        static inline MixedCase* UPPERCASE(ActivationEventInfo *a) {            \
+                if (_unlikely_(!a || a->trigger_unit_type != UNIT_##UNIT_TYPE)) \
+                        return NULL;                                            \
+                                                                                \
+                return (MixedCase*) a;                                          \
+        }
+
+/* For casting the various unit types into a unit */
+#define ACTIVATION_EVENT_INFO(u)                                      \
+        ({                                                            \
+                typeof(u) _u_ = (u);                                  \
+                ActivationEventInfo *_w_ = _u_ ? &(_u_)->meta : NULL; \
+                _w_;                                                  \
+        })
+
+ActivationEventInfo *activation_event_info_new(Unit *trigger_unit);
+ActivationEventInfo *activation_event_info_ref(ActivationEventInfo *p);
+ActivationEventInfo *activation_event_info_unref(ActivationEventInfo *p);
+void activation_event_info_serialize(ActivationEventInfo *p, FILE *f);
+int activation_event_info_deserialize(const char *key, const char *value, ActivationEventInfo **info);
+int activation_event_info_append_env(ActivationEventInfo *info, char ***strv);
+int activation_event_info_append_pair(ActivationEventInfo *info, char ***strv);
+DEFINE_TRIVIAL_CLEANUP_FUNC(ActivationEventInfo*, activation_event_info_unref);
+
+typedef struct ActivationEventInfoVTable {
```

Comment:
why have a separate vtable for this? why not use the unit vtable? you only have one of these per unit type, so it shoud just be the same?

Reply:
because the lifetime and management of the info object is independent from the lifetime and management of the unit object (given they are also owned by jobs), so it seems much cleaner and tidier to keep it separate, conceptually, even if in practice it wouldn't make a difference